### DISCUSSION TITLE
Why the css will repeat ?

### DISCUSSION THREAD

**Discussion by xjArea** [vikejs/vike#1003] [Discussion] [answered]
😰 During the development stage, I found that the CSS was repeated, once from the `<link>` tag, and once from the inner `<style>` tag, and they all took effect, which caused us to see repeated styles in the devTool panel, This is definitely not a good thing, because the repeated css code will cause me a lot of trouble when debugging.

Although this problem does not exist after build, I can't understand why this happens here. Generally speaking, since the code has been introduced with `<link>`, there is no need to add it as an internal `<style>`, what is wrong here, is there a way we can avoid it?

Here is a screenshot of the css repetition:  
![image](https://github.com/brillout/vite-plugin-ssr/assets/35623539/7dd6cbab-a99c-49fe-9c79-f9619ba79179)

**Comment by brillout** [comment]
The CSS `<link>` tag is injected by Vike to avoid [FOUC](https://en.wikipedia.org/wiki/Flash_of_unstyled_content), while the `<style>` tags are automatically injected by Vite (and Vike doesn't have control over these ones).

~~I think we can remove `<link>` once the page has loaded.~~ I'm actually realizing that, because of Vite's on-demand transpiling (and therefore on-demand style injecting), there isn't an easy way to tell when Vite is done injecting `<style>` tags.

**Comment by xjArea** [comment]
Theoretically speaking, between the two, I prefer to delete the `<style>` tag, because it will make the page more concise, and it is easier to know which file it comes from when looking for `selector`, but If this is difficult to achieve at the code level, delete the `<link>` is OK. 😊

**Reply by brillout** [reply]
I'd rather not interfere with Vite's style handling.

**Reply by xjArea** [reply]
If `<link>` is located in `<head>`, then there should be no FOUC, because `<body>` will not be rendered until `<link>` is loaded and executed, and the .css file There is also a blocking effect at the rendering level. emmm, it should be like this. 😅

**Reply by xjArea** [reply]
Alternatively, provide a parameter that allows the user to choose…

**Reply by brillout** [reply]
That's why precisely why VPS injects the `<link>` one.

**Comment by snake-py** [comment]
So is there away to avoid this? We have quite a lot of styles which already are quite demanding in dev.

**Reply by brillout** [reply]
The problem is [this](https://github.com/brillout/vite-plugin-ssr/discussions/1003#discussioncomment-6370732), so I can't think much of a solution.

But I'm happy to further dig depending on your use case for it.

> We have quite a lot of styles which already are quite demanding in dev.

What's the issue exactly?

**Reply by snake-py** [reply]
To be honest I would not care in dev if there would be FOUC, if the app would start up faster. I guess the main concern is that we right now are up to minute till the app if we have some scss changes till the reload propergates. But I don't think that this a VPS concern, we have been seeing this issue in other apps as well seems to be more of a scss issue. Also our Dev console is super super slow to the poitn that it becomes unusable to do some style changes.

**Reply by snake-py** [reply]
@brillout I just confirmed that the slow behaving dev tools is caused that all styles are injeted two times. So I really need an option to somehow opt out of this behaivour for dev, even if it means to get FOUC.

**Reply by snake-py** [reply]
In fact only keeping the link injection will work, it seems like that the style tags slow the dev tools down. Deleting the style tags from the dom massively enhances the dev tools.

**Reply by snake-py** [reply]
Basically this solves my issue, but this seems way off a hack to do it. 

```tsx
  useEffect(() => {
    if (import.meta.env.DEV) {
      document.querySelectorAll('style').forEach((style) => {
        style.remove();
      });
    }
  }, []);
```

**Reply by brillout** [reply]
Vike only injects `<link>` tags, so it's either a SCSS or Vite thing and I'm not sure there is much Vike can do about.

Would migrating away from SCSS be an option for you? For example to PostCSS, which Vite has built-in support for. And Vite's support for [Lightning CSS](https://lightningcss.dev/) is also very promising.

**Reply by brillout** [reply]
That said, there could be a Vike setting to omit injecting `<link>` tags, but seems like what you want is getting rid of the `<style>` tags.

**Reply by snake-py** [reply]
> That said, there could be a Vike setting to omit injecting <link> tags, but seems like what you want is getting rid of the <style> tags.

Exactly, I need to get rid of the style tags.

> Would migrating away from SCSS be an option for you? 

I would actually love to migrate away from scss, however our internal library relies heavily on prime react, and they work with scss. They seem to be in the process of migrating to common css. Once their migration is finish we want to migrate towards css modules or css-in-js.

> support for [Lightning CSS](https://lightningcss.dev/)

First time seeing this will look into that.

**Reply by snake-py** [reply]
I think for now I will just leave my wacky work around and need to wait that prime will update their lib, so we can start to migrate away from scss.

**Reply by brillout** [reply]
:+1:

**Comment by 7iomka** [comment]
@brillout 
Hasn't anyone solved this problem yet?
I hate to work with css with dublicates, when I need to disable some styles immediatelly, I need to disable twice is dev tools...

with
```
css: {
    devSourcemap: true,
  },
```
I have the same filename, but the wrong path to it, and because of this I this I still have dublicate
![image](https://github.com/vikejs/vike/assets/18473137/cc610dc3-8ce3-43b6-9f11-c93436c53657)

Any suggesion for a hack to somehow disable dublicates in dev?

**Reply by brillout** [reply]
@7iomka The root issue is explained [here](https://github.com/vikejs/vike/discussions/1003#discussioncomment-6370732). I'd love to remove the `<link>` tag injected by Vike, but how can we tell that Vite has finished loading the assets of the initial page?

**Reply by brillout** [reply, marked as answer]
[Fix](https://github.com/vikejs/vike/commit/fae90a15d88e5e87ca9fcbb54cf2dc8773d2f229) pre-released as `0.4.178-commit-fae90a1`.

Btw. would your company be up for [sponsoring](https://github.com/vikejs/vike/issues/1350)? It makes a massive difference to us :green_heart:
